File: api/schemas/user_sessions.py
from datetime import datetime
from typing import Optional

from api.schemas.base import ORMBase


class SessionSignOutRequest(ORMBase):
    sign_out_reason_name: str
    fingerprint_hash: Optional[str] = None


class ExpiredSessionSignOutRequest(SessionSignOutRequest):
    session_identifier: str


class UserSessionSummary(ORMBase):
    session_identifier: str
    device_label: str | None = None
    device_type: str | None = None
    browser: str | None = None
    operating_system: str | None = None
    ip_address: str | None = None
    signed_in_at: datetime
    last_seen_at: datetime
    signed_out_at: datetime | None = None
    is_active: bool
    sign_out_reason_name: str | None = None
    is_current: bool


class KnownDeviceSummary(ORMBase):
    device_key: str
    device_label: str | None = None
    device_type: str | None = None
    browser: str | None = None
    operating_system: str | None = None
    session_count: int
    active_session_count: int
    signed_in_at_first: datetime
    last_seen_at: datetime
    is_current_device: bool


class UserSessionsResponse(ORMBase):
    current_session_identifier: str | None = None
    sessions: list[UserSessionSummary]
    known_devices: list[KnownDeviceSummary]


class AdminUserSessionSummary(UserSessionSummary):
    user_id: int
    username: str
    full_name: str | None = None
    display_name: str | None = None
    role_name: str | None = None


class AdminActiveUserSessionsResponse(ORMBase):
    active_user_count: int
    active_session_count: int
    sessions: list[AdminUserSessionSummary]


class CurrentSessionStatusResponse(ORMBase):
    session_identifier: str
    is_active: bool

File: frontend/src/interfaces/UserSessions.ts
export interface UserSessionSummary {
  session_identifier: string;
  device_label?: string | null;
  device_type?: string | null;
  browser?: string | null;
  operating_system?: string | null;
  ip_address?: string | null;
  signed_in_at: string;
  last_seen_at: string;
  signed_out_at?: string | null;
  is_active: boolean;
  sign_out_reason_name?: string | null;
  is_current: boolean;
}

export interface KnownDeviceSummary {
  device_key: string;
  device_label?: string | null;
  device_type?: string | null;
  browser?: string | null;
  operating_system?: string | null;
  session_count: number;
  active_session_count: number;
  signed_in_at_first: string;
  last_seen_at: string;
  is_current_device: boolean;
}

export interface UserSessionsResponse {
  current_session_identifier?: string | null;
  sessions: UserSessionSummary[];
  known_devices: KnownDeviceSummary[];
}

export interface AdminUserSessionSummary extends UserSessionSummary {
  user_id: number;
  username: string;
  full_name?: string | null;
  display_name?: string | null;
  role_name?: string | null;
}

export interface AdminActiveUserSessionsResponse {
  active_user_count: number;
  active_session_count: number;
  sessions: AdminUserSessionSummary[];
}
